src/main/java/org/jboss/aerogear/keycloak/metrics/ResourceExtractor.java

```diff
@@ -0,0 +1,57 @@
+package org.jboss.aerogear.keycloak.metrics;
+
+import org.jboss.logging.Logger;
+
+import javax.ws.rs.core.UriInfo;
+import java.util.List;
+
+class ResourceExtractor {
+
+    private final static Logger logger = Logger.getLogger(ResourceExtractor.class);
+
+    private ResourceExtractor() {
+    }
+
+    /**
+     * This method parses a list of path segments (typically obtained from {@link UriInfo}) and
+     * returns a single string with a resource name without information about the realm. For
+     * the static resources, this method returns an empty string.
+     *
+     * @param pathSegments Path Segments - usually obtained from {@link UriInfo}.
+     * @return The resource name.
+     */
+    static String getResource(List<String> pathSegments) {
+        try {
+            StringBuilder ret = new StringBuilder();
+            int realmIndex = -1;
+            boolean isAdminEndpoint = false;
+            String path = null;
+            for(int i = 0; i < pathSegments.size(); ++i) {
+                path = pathSegments.get(i);
+                if ("realms".equals(path)) {
+                    realmIndex = i;
+                }
+                if (i == 0 && "admin".equals(path)) {
+                    isAdminEndpoint = true;
+                }
+                if (isAdminEndpoint && i > realmIndex + 2) {
+                    // We're parsing admin endpoint and we're not interested in the
+                    // resource ID.
+                    break;
+                }
+                if (realmIndex == -1 || i > realmIndex + 1) {
```

Comment:
This seems to append everything to the string except the segment directly following the realm. How does it protect us from adding user ids to the result string?

The crucial thing here is to avoid adding information with lots of variation (like ids), because it will blow up the metrics size.

Reply:
I think you're right @pb82. Looking at our REST model (https://www.keycloak.org/docs-api/12.0/rest-api/index.html#_clients_resource) it seems we put most of the IDs (either client, user, flow or any other resource) in path params. I probably need another approach to get rid of them.

Reply:
Analyzing Matched URIs seems promissing:

```
13:28:52,766 INFO  [stdout] (default task-2) Matched URIs: [resources/89pe1/admin/logo-example/partials/client-list.html, resources]
13:30:35,495 INFO  [stdout] (default task-3) Matched URIs: [realms/master/account, realms/master/account, realms]
13:32:53,160 INFO  [stdout] (default task-10) Matched URIs: [realms/master/.well-known/openid-configuration, realms]
13:33:16,656 INFO  [stdout] (default task-10) Matched URIs: [realms/master/protocol/openid-connect/auth, realms/master/protocol/openid-connect, realms]
13:33:47,319 INFO  [stdout] (default task-10) Matched URIs: [realms/master/protocol/openid-connect/certs, realms/master/protocol/openid-connect, realms]
13:34:37,768 INFO  [stdout] (default task-10) Matched URIs: [realms/master/account/applications, realms/master/account, realms]
13:35:48,003 INFO  [stdout] (default task-11) Matched URIs: [admin/realms/master/client-scopes/ee5e5908-a0b6-43c8-b213-fb452f129a5e, admin/realms/master/client-scopes, admin/realms/master, admin/realms, admin]
13:36:24,642 INFO  [stdout] (default task-11) Matched URIs: [admin/realms/master/users/171753bc-8184-4989-929b-288fdc661b90, admin/realms/master/users, admin/realms/master, admin/realms, admin]
13:36:24,793 INFO  [stdout] (default task-11) Matched URIs: [admin/realms/master/attack-detection/brute-force/users/171753bc-8184-4989-929b-288fdc661b90, admin/realms/master/attack-detection, admin/realms/master, admin/realms, admin]
```

So I think the algorithm might take just the last two components from the vector. The only `if` that we'd have would be to "resources". If the resources are detected to be the last element - we'd just take the last item.

Reply:
@pb82 @maleck13 WDYT about this approach? Would it give us enough granularity?

Reply:
Sorry. I am not following what the metric would look like? 

In the case of 
```
[realms/master/protocol/openid-connect/certs, realms/master/protocol/openid-connect, realms]
```
would it look like

```
keycloak_request_duration_bucket{resource="realms/master/protocol/openid-connect/realms", method="POST"...}

```
Or something like 

```
keycloak_request_duration_bucket{resource="realms/master/protocol/openid-connect/", otherKey="realms", method="POST"...}

```

Reply:
Here's the output from my testing environment:

```
keycloak_request_duration_bucket{method="GET",resource="realms,realms/master/protocol/openid-connect",le="50.0",} 1.0
...
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="50.0",} 0.0
...
```

Reply:
wondering if splitting into two values would make it simpler to deal with from a query point of view. Might be a separate PR but might also be helpful to group by code.

```
keycloak_request_duration_bucket{method="GET",resource="realms", endpoint="realms/master/protocol/openid-connect",le="50.0",} 1.0
```

Reply:
That would be tricky as 99% of the request will fall into "realms" resource - everything that starts with `https://$KC/auth/realms`. Unfortunately this is the way the Keycloak JAX-RS resources are structured. The only one thing that you'd have separate is the admin endpoint. 

To sum it up - because of the way we use JAX-RS in Keycloak, it's very hard to do such things without modifying the code on the server (and attaching an SPI implementation).

Reply:
ok do we have any idea how many different values this would create? I guess it depends on the number of rest endpoints? Otherwise I think this is fine. Might be able to graph by doing a group by resource and also add specific alert around service accounts or openid-connect endpoint

Reply:
Yeah, it's hard to test all the endpoints at once. But in general you should have a distinction between the admin endpoints and the protocol ones (which is crucial for us).